drivers/serial/uart_stm32.c

```diff
@@ -1446,6 +1447,9 @@ static int uart_stm32_pm_control(const struct device *dev,
 		LL_USART_ClearFlag_ORE(UartInstance);
 		/* Leave UartInstance unchanged */
 		break;
+	case PM_DEVICE_ACTION_RESUME:
+		/* Leave UartInstance ready to Tx/Rx */
+		break;
```

Comment:
```suggestion
```

After checking, this should not be required, since the system doesn't check for any sort of error when resuming (should likely be fixed...) https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/pm/device.c#L72

Reply:
Even if system doesn't check for errors, this actually has some subtle impacts as just adding this line is sufficient to get the test pass with twister (I though it could be ASSERT impact, but after some test, it doesn't seem so).

Reply:
Just reproduced now, I'm now a bit worried. Let me investigate this a bit more, it should not happen, I think.

Reply:
Running the test as a regular application and monitoring the serial port produces this:

- Adding `case PM_DEVICE_ACTION_RESUME:`
```
PROJECT EXECUTION SUCCESSFUL
```
- Removing `case PM_DEVICE_ACTION_RESUME:`
```
...
fROJECT EXECUTION SUCCESSFUL
```
^ note the `f` here.

Behavior should be the same, since the action does nothing and system PM does not check for errors when a device is resumed (so returning -ENOTSUP should result in the same behavior).

Reply:
> If `PM_DEVICE_ACTION_LOW_POWER` and `PM_DEVICE_ACTION_SUSPEND` actions return `-ENOTSUP`

What would be the difference with completely removing `pm_control` hook from system PM point of view ?
If the result of system PM evolution is that current implementation is not more in line/useful, it may make
more sense to remove it totally (for this LTS).
Then, we would start from a clean sheet for a proper implementation starting V3.0.0.

Reply:
I think we're facing 2 problems here:

- STM32 UART driver needs to block entering LP mode if UART is active. This can be solved using constraints in some cases (IRQ/DMA) but not in others (polling). In such case, the _best_ driver can do is to block the transition when the PM hook is called (note that this means that `CONFIG_PM_DEVICE` is a de-facto requirement if `CONFIG_PM=y`). The fact that `PM_DEVICE_ACTION_LOW_POWER` is used by certain PM states means that this case needs to be handled, not because the device needs to enter "low power" but because it also needs to consider this case.
- If `PM_DEVICE_ACTION_RESUME` is handled (return 0) the test seems to execute correctly. If not present, or present but returning `-ENOTSUP`, the test fails to pass (there seems to be some garbage in the UART output). This looks like a separate problem, since the return code issued by the resume operating while the system wakes up is not relevant, doesn't change the behavior as far as I can see.